### PR TITLE
Build Threading.Tasks.DataFlow and ComponentModel.Annotations for NetCoreAppCurrent

### LINKED ISSUE #48662: Threading.Tasks.DataFlow and ComponentModel.Annotations are not building NetCoreAppCurrent
These libraries are included in the shared framework, but we aren't building them for `$(NetCoreAppCurrent)`.

https://github.com/dotnet/runtime/blob/8a3fd5a065b019cc70a600a759ea0c87c14cae8f/src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj#L3

https://github.com/dotnet/runtime/blob/8a3fd5a065b019cc70a600a759ea0c87c14cae8f/src/libraries/System.ComponentModel.Annotations/src/System.ComponentModel.Annotations.csproj#L3

This causes the following problems:

1. These libraries have `<Nullable>enable</Nullable>` in them, but they are building against netstandard APIs, which don't have nullablility annotations.
2. When someone wants to use a new API (or new attributes, like `DynamicDependency` and `UnconditionalSuppressMessage`), they are not available.

We should add a `$(NetCoreAppCurrent)` target for these libraries, and ship that in the shared framework / runtimepack. I analyzed a recent shared framework, and these were the only 2 libraries (outside of `System.Runtime.CompilerServices.Unsafe`) that didn't target `net6.0`.

cc @LakshanF @joperezr @safern @ViktorHofer @ericstj

### PR DESCRIPTION
**Pull request by eerhardt**
Fix #48662

This involved fixing a few nullable annotation errors in System.Threading.Tasks.DataFlow.

cc @ericstj

### CODE REVIEW

**Comment by ericstj** [comment]
@ViktorHofer and I were just discussing this the other day. 

There's no requirement to target `$(NetCoreAppCurrent)` just because some assembly is part of the shared framework.  In fact we try to avoid configurations that don't ship in the package since they create a risk for a package downgrading functionality.

What benefit are we getting by targeting `NetCoreAppCurrent`?  It looks like nullable annotations?  Could we get the same benefit by ensuring the references are nullable annotated?

**Comment by eerhardt** [comment]
> What benefit are we getting by targeting NetCoreAppCurrent?

1. The nullable annotations working correctly.
2. @LakshanF was annotating System.Threading.Tasks.DataFlow for linker annotations and needed to reference `UnconditionalSuppressMessageAttribute`, which is only in `net5.0`+. So he added the file as a source linked file, which compiles the attribute as `internal`. This then broke the linker because it was picking up the `internal` version of `UnconditionalSuppressMessageAttribute` and erroring with: 

```
D:\work\Core\changes\runtime\src\libraries\System.Private.CoreLib\src\ILLink\ILLink.Suppressions.Shared.xml : error IL20
52: Property 'Scope' could not be found [D:\work\Core\changes\runtime\src\libraries\src.proj]
D:\work\Core\changes\runtime\src\libraries\System.Private.CoreLib\src\ILLink\ILLink.Suppressions.Shared.xml : error IL20
52: Property 'Target' could not be found [D:\work\Core\changes\runtime\src\libraries\src.proj]
```

Having to compile `internal` attributes in the shared framework assemblies seems busted. Any shared framework assembly that needs to use those attributes should be able to pick those attributes up from the public API.

cc @vitek-karas

**Comment by safern** [comment]
> Could we get the same benefit by ensuring the references are nullable annotated?

We don't get the same benefit as .NET Standard and .NET Framework are not annotated, so you don't get data flow analysis on stuff like `Debug.Assert` or `string.IsNullOrEmpty` nor any of the APIs that affect data flow analysis that are inbox on those frameworks as their ref assemblies are not annotated. This leads annotating libraries very hard and basically it is just throwing a bunch of `!`s across the code. That is why we decided to disable nullable warnings for anything non netcoreapp.

**Comment by ericstj** [comment]
That's a bummer.  So long as we have these configurations we should be really careful about them being noticeably different from the package.   At the moment we only have code review to catch significant differences.

File: src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj [hunk 1]
```diff
@@ -60,7 +61,16 @@
...
+  <ItemGroup Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">
+    <Reference Include="System.Collections" />
+    <Reference Include="System.Collections.Concurrent" />
+    <Reference Include="System.Linq" />
+    <Reference Include="System.Runtime" />
+    <Reference Include="System.Threading" />
+    <Reference Include="System.Threading.Thread" />
+    <Reference Include="System.Threading.ThreadPool" />
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard1.0' or '$(TargetFramework)' == 'netstandard1.1' or $(TargetFramework.StartsWith('net4'))">
```
Comment (ViktorHofer) [inline src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj:73]: nit, if the tfm isn't rid specific, just use equality checks like for netstandard.

```suggestion
  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard1.0' or '$(TargetFramework)' == 'netstandard1.1' or '$(TargetFramework)' == 'net461'">
```

**Review by safern — APPROVED** [review]
Looks reasonable

File: src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj (hunk 1 above)
Comment (eerhardt) [reply src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj:73]: I copied that from the top of this file:

```xml
<DefineConstants Condition="'$(TargetFramework)' == 'netstandard1.1' or '$(TargetFramework)' == 'netstandard2.0' or $(TargetFramework.StartsWith('net4'))">$(DefineConstants);FEATURE_TRACING</DefineConstants>
```

Should I fix that one as well, if that's how the condition should be written?

**Review by stephentoub — APPROVED** [review]
Dataflow nullable annotation changes LGTM

File: src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj (hunk 1 above)
Comment (ViktorHofer) [reply src/libraries/System.Threading.Tasks.Dataflow/src/System.Threading.Tasks.Dataflow.csproj:73]: I'm fine with either. I will go over the libraries projects at some point anyway and clean-up the conditions. The new msbuild property functions that shipped with NET5 allow to condition on the TargetFrameworkIdentifier, TargetFrameworkPlatform, etc. directly in the project file which will simplify most of our conditions.

**Comment by eerhardt** [comment]
Looks like I missed a `DefineConstants` when building for netcoreapp. (Thanks for the catch, unit tests!)

Not a huge change, but if anyone wants to take another look here.

**Comment by eerhardt** [comment]
Failures are https://github.com/dotnet/runtime/issues/1488

**Pull request merged**